test/jdk/java/util/Objects/BasicObjectsTest.java

```diff
@@ -204,106 +183,171 @@ private static int testRequireNonNull() {
-        } catch (NullPointerException npe) {
-            // expected
-            errors += npe.getMessage().equals("supplier.get()") ? 0 : 1;
-        }
-        try {       // supplier is null
-            Objects.requireNonNullElseGet(null, null);
-            errors += 1;
-        } catch (NullPointerException npe) {
-            // expected
-            errors += npe.getMessage().equals("supplier") ? 0 : 1;
-        }
-        return errors;
+        assertSame(defString, Objects.requireNonNullElseGet(nullString, () -> defString),
+            "requireNonNullElseGet(null, supplier) should return supplier result");
+
+        assertSame(nonNullString, Objects.requireNonNullElseGet(nonNullString, () -> defString),
+            "requireNonNullElseGet(non-null, supplier) should return non-null");
+
+        assertSame(nonNullString, Objects.requireNonNullElseGet(nonNullString, () -> null),
+            "requireNonNullElseGet(non-null, null-supplier) should return non-null");
+
+        NullPointerException npe2 = assertThrows(
+            NullPointerException.class,
+            () -> Objects.requireNonNullElseGet(null, () -> null),
+            "requireNonNullElseGet(null, null-returning supplier) should throw NPE");
+        assertEquals("supplier.get()", npe2.getMessage(),
+            "Exception message should be 'supplier.get()'");
+
+        NullPointerException npe3 = assertThrows(
+            NullPointerException.class,
+            () -> Objects.requireNonNullElseGet(null, null),
+            "requireNonNullElseGet(null, null supplier) should throw NPE");
+        assertEquals("supplier", npe3.getMessage(),
+            "Exception message should be 'supplier'");
+    }
+
+    /**
+     * Test requireNonNull with null Supplier parameter.
+     * Should throw NullPointerException with null message (as per API spec).
+     */
+    @Test
+    public void testRequireNonNullWithNullSupplier() {
+        // Test with null object and null supplier
+        NullPointerException npe = assertThrows(
+            NullPointerException.class,
+            () -> Objects.requireNonNull(null, (Supplier<String>) null),
+            "requireNonNull(null, null supplier) should throw NPE");
+        // According to Objects.java implementation, when messageSupplier is null,
+        // the exception message is also null
+        assertNull(npe.getMessage(),
+            "Exception message should be null when messageSupplier is null");
+    }
+
+    /**
+     * Test requireNonNull with Supplier that throws exception.
+     * The exception from supplier should be thrown, not wrapped.
+     */
+    @Test
+    public void testRequireNonNullWithSupplierThrowingException() {
+        // The exception from supplier should be thrown directly, not wrapped
+        RuntimeException re = assertThrows(
+            RuntimeException.class,
+            () -> Objects.requireNonNull(null, () -> {
+                throw new RuntimeException("Supplier exception");
```

Comment:
Instead of detecting by message, you can allocate an exception outside of the lambda, throw it here, and assertSame your exception with the assertThrows returned exception.